scanner/decisions/0005-scannerv4-read-cvss-data.md

```diff
@@ -0,0 +1,44 @@
+```
+curl "https://services.nvd.nist.gov/rest/json/cves/2.0?cvssV3Severity=CRITICAL&startIndex=0" > critical-1.json
+``` 
+For every severity level, the workflow iteratively exhaust the startIndex to capture all associated data. And there are four severity levels: LOW, MEDIUM, HIGH, CRITICAL. After all json files downloaded, they will be compress to one data bundle as one single compressed file.
+
+We've chosen to base our API requests on CVSS V3 severity to ensure all data integrated into Claircore possesses valid CVSS V3 metrics. Consequently, retrieving data without CVSS V3 information becomes irrelevant.
+
+The CVSS Updater in Central will be modified to download compressed file (generated by CI mentioned above) in the Google bucket. This eliminates the need for JSON parsing in the updater. Scanner V4 will pull the data bundle from Central and populate the Matcher DB with the enrichment data.
+
+A custom enricher is essential for Scanner V4, as it will provide a mechanism to retrieve JSON data from Central, parse the enrichment details, and store them in the Scanner V4 database.  Additionally, our data will be sourced from the NVD CVE API rather than JSON feeds, so the custom enricher can accommodate the distinct data formats. This also offers greater flexibility, allowing us to extract more information from the fetched JSONs than what Claircore offers.
+
+The CVSS Updater in Central will attempt to download the data bundle from the Google bucket using exponential retries in case of partial failures, up to a maximum interval within the next loop, which happens every 4 hours by default.
+
+Should the GitHub Actions workflow (CI) experience difficulties in accessing or downloading segments of the CVSS data due to an HTTP request failure—potentially caused by the NVD API being unavailable or due to rate limiting—the affected HTTP request should be automatically retried following an exponential backoff strategy. In the event that all retry attempts fail, the workflow will be terminated, and the system will await the next scheduled update in 4 hours. The team will be promptly notified of any aborted or failed workflows via Slack.
+
+## Consequences
+
+One drawback of the NVD CVE API is its rate limit: only 5 requests are allowed per 30 seconds without an API Key. This restriction necessitates that our workflow fetch data at a slower pace. Additionally, transitioning from the NVD JSON feed to the NVD CVE API brings in pagination limitations. For example, when fetching data with a "low" severity, the API provides a maximum of 2000 CVEs per response. This means multiple requests are required to retrieve the complete dataset for "low" severity. The result is an increased number of smaller JSON files. Consequently, we need to compress all JSON files into a single archived file such as tar.gz or tar.xz.
```

Comment:
Another option is to buffer the vulnerabilities until we hit maybe 20,000 (random number) and then write them to a file. Have we considered this?

Reply:
Buffering in memory? Any benefit to the producer or consumer compared to using files?

Reply:
We probably don't need to decompress that tar file so we should be good.

Reply:
I expect we will decompress the tar file when parsing the JSON to insert it into the DB. My question was, what is the benefit of buffering in memory to create larger files vs. using smaller files?

Reply:
I discussed with @RTann . He asked the question because he was concerned there will be many json files to be decompressed in scanner. But for the gzreader we can actually iterate the files in tar. @jvdm